libs/primitives/src/conversion.rs

```diff
@@ -0,0 +1,88 @@
+use sp_arithmetic::{
+	traits::{ensure_pow, BaseArithmetic},
+	ArithmeticError, FixedPointNumber,
+};
+
+/// Transform a fixed point number to a Balance.
+/// The resulting Balance will be represented with the `decimals` given.
+pub fn fixed_point_to_balance<
+	FixedPoint: FixedPointNumber<Inner = IntoBalance>,
+	IntoBalance: BaseArithmetic + Copy,
+>(
+	fixed_point: FixedPoint,
+	decimals: usize,
+) -> Result<IntoBalance, ArithmeticError> {
+	let integer_part = fixed_point.into_inner().ensure_div(FixedPoint::DIV)?;
+	let frac_part = fixed_point
+		.into_inner()
+		.ensure_sub(integer_part * FixedPoint::DIV)?;
+
+	let magnitude = ensure_pow(IntoBalance::from(10), decimals)?;
+
+	let new_integer_part = integer_part.ensure_mul(magnitude)?;
+	let new_frac_part = if magnitude > FixedPoint::DIV {
+		frac_part.ensure_mul(magnitude.ensure_div(FixedPoint::DIV)?)?
+	} else {
+		frac_part
+			.ensure_mul(magnitude)?
+			.ensure_div(FixedPoint::DIV)?
+	};
+
+	new_integer_part.ensure_add(new_frac_part)
+}
+
+#[cfg(test)]
+mod tests {
+	use frame_support::{assert_err, assert_ok};
+	use sp_arithmetic::fixed_point::FixedU64;
+
+	use super::*;
+
+	#[test]
+	fn with_no_decimals() {
+		assert_ok!(fixed_point_to_balance(FixedU64::from_float(23.42), 0), 23);
+	}
+
+	#[test]
+	fn with_less_decimals_than_div() {
+		assert_ok!(
+			fixed_point_to_balance(FixedU64::from_float(23.42), 6),
+			23_420_000
+		);
+
+		assert_ok!(fixed_point_to_balance(FixedU64::from_float(23.42), 0), 23);
+	}
+
+	#[test]
+	fn with_same_decimals_as_div() {
+		assert_ok!(
+			fixed_point_to_balance(FixedU64::from_float(23.42), 9),
+			23_420_000_000
+		);
+	}
+
+	#[test]
+	fn with_more_decimals_than_div() {
+		assert_ok!(
+			fixed_point_to_balance(FixedU64::from_float(23.42), 12),
+			23_420_000_000_000
+		);
+	}
+
+	#[test]
+	fn with_max_decimals() {
+		assert_ok!(
+			fixed_point_to_balance(FixedU64::from_float(23.42), 17),
```

Comment:
That `17` limitation comes from both, the number used for this example and the `FixedU64`, which can only hold numbers around 10^18. Using `FixedU128` and `u128` as `Balance`, which is what we should use in our runtime, we can have 18 decimals and 18 zeros for the integer part, which is Ok for all currencies.

Actually, I think this test is not testing anything at all and can be removed to avoid confusions

Reply:
Thanks for the clarification! :)